### PR TITLE
[Train] Add Intel HPU Support and Training Scripts for BERT and ResNet in Ray Train

### PR DESCRIPTION
**Pull request by zhangjian94cn**
## Why are these changes needed?

To leverage the potential of Intel's Habana Processing Units (HPUs), we extend Ray Train's capabilities by adding support for Intel HPU hardware. This update also includes training scripts for BERT and ResNet models, which are widely used in NLP and computer vision tasks, respectively.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/torch/config.py [hunk 1]
```diff
@@ -14,6 +14,11 @@
 from ray.train.constants import DEFAULT_NCCL_SOCKET_IFNAME
 from ray.util import PublicAPI
 
+from ray._private.accelerators.hpu import HPU_PACKAGE_AVAILABLE
+if HPU_PACKAGE_AVAILABLE:
+    import habana_frameworks.torch.core as htcore
+    import habana_frameworks.torch.distributed.hccl as hpu_dist
```
Comment (woshiyyya) [inline python/ray/train/torch/config.py:20]: Where did you use `htcore` and `hpu_dist` after import?  Any environment variable need to be set for hccl backend?

File: python/ray/air/_internal/torch_utils.py [hunk 2]
```diff
@@ -56,6 +59,8 @@ def get_device() -> Union[torch.device, List[torch.device]]:
 
         devices = [torch.device(f"cuda:{device_id}") for device_id in device_ids]
         device = devices[0] if len(devices) == 1 else devices
+    elif HPU_PACKAGE_AVAILABLE and torch_hpu.is_available():
+        device = torch.device("hpu")
```
Comment (woshiyyya) [inline python/ray/air/_internal/torch_utils.py:63]: Is it possible to allocate multiple hpu devices for a worker?

File: doc/source/train/examples/hpu/bert.ipynb [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+    "\n",
+    "    train_config = {\n",
+    "        \"lr\": 1e-3,\n",
+    "        \"epochs\": 10,\n",
+    "        \"batch_size_per_worker\": global_batch_size // num_workers,\n",
+    "    }\n",
+    "\n",
+    "    # Configure computation resources\n",
+    "    scaling_config = ScalingConfig(num_workers=num_workers, resources_per_worker={\"CPU\": 1, \"HPU\": 1})\n",
+    "    torch_config = TorchConfig(backend = \"hccl\")\n",
```
Comment (woshiyyya) [inline doc/source/train/examples/hpu/bert.ipynb:185]: Can you implement the hccl backend option in `TorchConfig`?

File: doc/source/train/examples/hpu/resnet.ipynb [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def download_datasets():\n",
+    "    os.system(\"wget https://download.pytorch.org/tutorial/hymenoptera_data.zip >/dev/null 2>&1\")\n",
+    "    os.system(\"unzip hymenoptera_data.zip >/dev/null 2>&1\")\n",
+    "download_datasets()"
```
Comment (woshiyyya) [inline doc/source/train/examples/hpu/resnet.ipynb:91]:
```suggestion
    "    os.system(\"unzip hymenoptera_data.zip >/dev/null 2>&1\")"
```
remove the download call here?

**Comment by woshiyyya** [comment]
@zhangjian94cn The code is pretty clean. Thanks for the contribution! 

Here are several improvements:

For the code change:
- Is there's any environment variables need to be set for the hccl backend?
- Fix the CI tests?

For examples:
- Can you add the cluster configuration (e.g. num hpu cores, num of nodes, instance type) at the beginning?
- Also can you include the required packages for users to run the example?
- Also can you run the notebook on hpu and keep the output cells? Just to confirm it works out of the box.

**Review by woshiyyya — CHANGES_REQUESTED** [review]

File: python/ray/train/torch/config.py (hunk 1 above)
Comment (kira-lin) [reply python/ray/train/torch/config.py]: htcore is imported to initialize the habana PyTorch backend; hccl is imported to initialize the habana distributed backend.  Environment variable is not required. This is documented [here](https://docs.habana.ai/en/latest/PyTorch/PyTorch_Model_Porting/Porting_PyTorch_Models_to_Gaudi.html?highlight=htcore#importing-habana-frameworks-torch-core)

File: python/ray/air/_internal/torch_utils.py (hunk 2 above)
Comment (kira-lin) [reply python/ray/air/_internal/torch_utils.py:65]: No, HPU do not support multiple devices for single process.

**Pull request closed without merging**